spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/PartTreeDatastoreQuery.java

```diff
@@ -223,8 +233,34 @@ else if (isCountingQuery && this.tree.isDelete()) {
 		}
 	}
 
+	private Supplier<Builder<?>> getQueryBuilderSupplier(Object[] parameters) {
+		Supplier<Builder<?>> queryBuilderSupplier;
+		queryBuilderSupplier = StructuredQuery::newEntityQueryBuilder;
+		ProjectionInformation projectionInformation =
+				this.projectionFactory.getProjectionInformation(this.queryMethod.getReturnedObjectType());
+
+		if (projectionInformation != null &&
+				projectionInformation.getType() != this.entityType
+				&& projectionInformation.isClosed()) {
+			ProjectionEntityQuery.Builder projectionEntityQueryBuilder = Query.newProjectionEntityQueryBuilder();
+			projectionInformation.getInputProperties().forEach(propertyDescriptor -> {
+				projectionEntityQueryBuilder.addProjection(mapToFieldName(propertyDescriptor));
+			});
+			queryBuilderSupplier = () -> projectionEntityQueryBuilder;
+		}
+		return queryBuilderSupplier;
+	}
+
+
+	private String mapToFieldName(PropertyDescriptor propertyDescriptor) {
+		String name = propertyDescriptor.getName();
+		DatastorePersistentProperty persistentProperty =
+				(DatastorePersistentProperty) this.datastorePersistentEntity.getPersistentProperty(name);
+		return persistentProperty.getFieldName();
+	}
+
 	private Slice executeSliceQuery(Object[] parameters) {
-		EntityQuery.Builder builder = StructuredQuery.newEntityQueryBuilder()
+		StructuredQuery.Builder builder = getQueryBuilderSupplier(parameters).get()
```

Comment:
This is optional but I am starting to get confused with the Supplier thing - maybe if you find you're always calling `getQueryBuilderSupplier(parameters).get()` here, you can take the opportunity to just have `getQueryBuilder()` return the query builder directly rather than a supplier and then call `.get()`.

But if it's gonna be a big refactor job then just ignore this comment.